Fix issues and react warnings for verification code components

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Closes #1002, #1003, #1004

- https://github.com/woocommerce/google-listings-and-ads/commit/305c21c4e0fa92ea25bd3f301a7be3b1bfe3777d - Fix that the first verification code input isn't cleared after reset.
- https://github.com/woocommerce/google-listings-and-ads/commit/2188f6075aac882330a7f327a5ed10d5325fac9d - Fix that backspace doson't remove selected digit in the `<VerificationCodeControl>`.
- https://github.com/woocommerce/google-listings-and-ads/commit/49371b518e96d4babc1d135d26e14ff06a9d55e0 - Fix the warning of updating state on an unmounted component.
- https://github.com/woocommerce/google-listings-and-ads/commit/8f8734125bfa81d4c4e3356315583cf6f14bbdc3 - Fix the warning of missing unique key in the `<AppButton>`.

### Additional notes

- [x] 📝  Revert https://github.com/woocommerce/google-listings-and-ads/commit/f605132381c0ff3494f8b4305c575381e2fa5c8c to removed `<DemoVerifyPhoneNumberCard>` before merging this PR.

### Screenshots:

#### #1002 - Clear input after reset

![Kapture 2021-09-10 at 15 19 51](https://user-images.githubusercontent.com/17420811/132823434-a88d92b4-2db5-4944-9427-335eab371e25.gif)

#### #1003 - Delete selected value when pressing backspace

![Kapture 2021-09-10 at 15 16 42](https://user-images.githubusercontent.com/17420811/132823451-c7a19e79-18eb-4188-9008-31e38d03cab0.gif)

#### #1004 - Fix react warnings

![Kapture 2021-09-10 at 15 54 21](https://user-images.githubusercontent.com/17420811/132823468-60bef990-09e7-451a-a317-930ed6e37f64.gif)

### Detailed test instructions:

1. Go to the Setting page and there's a demo component for testing
2. Follow the replication steps in each issue to check if they were fixed
    - #1002
    - #1003
    - #1004

### Changelog entry

eason9487 committed:
Add a temporary component to demonstrate the phone number verification card.

eason9487 committed:
Fix that backspace doson't remove selected digit in the VerificationCodeControl

Address https://github.com/woocommerce/google-listings-and-ads/issues/1003

eason9487 committed:
Fix that the first verification code input isn't cleared after resetting

Address https://github.com/woocommerce/google-listings-and-ads/issues/1002

eason9487 committed:
Fix the warning of updating state on an unmounted component

Address https://github.com/woocommerce/google-listings-and-ads/issues/1004

eason9487 committed:
Fix the warning of missing unique key in the AppButton

Address https://github.com/woocommerce/google-listings-and-ads/issues/1004

tomalec commented on js/src/hooks/useIsMounted.js:
```diff
@@ -0,0 +1,21 @@
+ * Returns a function to check whether the caller component is mounted or unmounted.
+ * Usually, it's used to avoid the warning - Can't perform a React state update on an unmounted component.
+ *
+ * @return {() => boolean} A function returns a boolean indicates the caller component is mounted or unmounted.
```
:nail_care: 
I'm not born with English/English native, but Grammarly suggests:
```suggestion
 * @return {() => boolean} A function returns a boolean that indicates the caller component is mounted or unmounted.
```

tomalec commented on js/src/hooks/useIsMounted.js:
```diff
@@ -0,0 +1,21 @@
+ *
+ * @return {() => boolean} A function returns a boolean indicates the caller component is mounted or unmounted.
+ */
+export default function useIsMounted() {
```
:grey_question: 
Looks like native [`Node.isConnected`](https://developer.mozilla.org/en-US/docs/Web/API/Node/isConnected). I'm surprised that React does not have it somewhat built-in and we have to hand-craft it? Do we do something not popular/standard here?

eason9487 replied:
> I'm surprised that React does not have it somewhat built-in and we have to hand-craft it?

It had had a similar API `isMounted`, but was deprecated. https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html

> Do we do something not popular/standard here?

It is difficult to say whether this practice is popular/standard or not because it's not really an issue or bug in most cases when seeing that warning.

`useIsMounted` is here mainly to avoid that scary warning appearing, but in fact there's no memory leak problem nor additional side effects.

The warning has quite a long background, mainly because some earlier global store/state management libraries based on the "Flux Architecture" had subscription and unsubscription mechanisms to be plugged in the class components. As an observer to get the global state changes, if a component forgets to unsubscribe after unmounted, it might let the global store keep the reference of the component itself or the method declared within the component, which may further lead to memory leaks.

However, with the implementation of something like react-redux, these subscription processes have been replaced by Higher-Order Components (HOC) and the memory leak problem is no longer common. After react hooks, these warnings are also almost false positives alarm. As it turns out, [this warning will be adjusted in the new version of react](https://github.com/facebook/react/pull/22114), and we will no longer need `useIsMounted` then.

tomalec committed:
Add few code comments to `VerificationCodeControl`

focus managing code.

tomalec commented on js/src/components/contact-information/phone-number-card/verification-code-control.js:
```diff
@@ -103,18 +103,38 @@ export default function VerificationCodeControl( {
 		onCodeChangeRef.current( toCallbackData( initDigits ) );
 	}, [ resetNeedle ] );
 
+	/**
+	 * Since the <InputControl> has an internal state management that always controls the actual `value` prop the <input>,
+	 * the <InputControl> is forced to be a controlled input.
+	 * When using it, it's always necessary to specify `value` prop from the below <AppInputControl>
+	 * to avoid the warning - A component is changing an uncontrolled input to be controlled.
+	 *
+	 * @see https://github.com/WordPress/gutenberg/blob/%40wordpress/components%4012.0.8/packages/components/src/input-control/input-field.js#L47-L68
+	 * @see https://github.com/WordPress/gutenberg/blob/%40wordpress/components%4012.0.8/packages/components/src/input-control/input-field.js#L115-L118
+	 *
+	 * But after specifying the `value` prop,
+	 * the synchronization of external and internal `value` state will depend on whether the input is focused.
+	 * It'd sync external to internal only if the input is not focused.
+	 * So here we await the `digits` is reset back to `initDigits` by above useEffect and sync to internal value,
+	 * then move the focus calling after the synchronization tick finished.
+	 *
+	 * @see https://github.com/WordPress/gutenberg/blob/%40wordpress/components%4012.0.8/packages/components/src/input-control/input-field.js#L73-L90
+	 *
+	 * And it's also used for the first time focus after this component being mounted.
+	 */
+	useEffect( () => {
+		if ( digits === initDigits ) {
+			maybeMoveFocus( 0 );
+		}
+	}, [ resetNeedle, digits ] );
```
👍 Thanks for the detailed explanations :) It makes it clear why the code is here not where it was before. 
:nail_care: However, I think we could also add a note of why we need it at all/what it does, for somebody who reads the file as is, not as a change to the previous state.

I made a branch with a small proposal https://github.com/woocommerce/google-listings-and-ads/compare/fix/1002-1003-1004-verification-code...fix/1002-1003-1004-verification-code-comments, feel free to merge it if you like it :)

eason9487 replied:
Merged. Thanks!

tomalec reviewed (approved):
Reviewed the code, tested locally on WP58 & WC57, LGTM.
Left few comments about cosmetics.

Thanks for the demo component, and detailed explanations.

eason9487 committed:
Correct English grammar

Address https://github.com/woocommerce/google-listings-and-ads/pull/1007#discussion_r708616020

Co-authored-by: Tomek Wytrębowicz <tomalecpub@gmail.com>

eason9487 committed:
Merge branch 'fix/1002-1003-1004-verification-code-comments' into fix/1002-1003-1004-verification-code

eason9487 committed:
Adjust JSDoc for VerificationCodeControl

eason9487 committed:
Revert "Add a temporary component to demonstrate the phone number verification card."

This reverts commit f605132381c0ff3494f8b4305c575381e2fa5c8c.
